File: Adaptive Threshold for Image/AdaptiveTreshold.cpp
#include "opencv2/imgproc.hpp"
#include "opencv2/imgcodecs.hpp"
#include "opencv2/highgui.hpp"
#include <opencv2/opencv.hpp>
#include "opencv2/core.hpp"

using namespace cv;

int main() {

  Mat original = imread("imagessudoku.jpg", IMREAD_GRAYSCALE);
  
  // Noise removal
  Mat noise_removed;
  medianBlur(original, noise_removed, 5);

  Mat output1;
  threshold(original, output1, 127, 255, THRESH_BINARY);

  Mat output2;
  adaptiveThreshold(original, output2, 255, ADAPTIVE_THRESH_MEAN_C, THRESH_BINARY, 11, 2);

  Mat output3;
  adaptiveThreshold(original, output3, 255, ADAPTIVE_THRESH_GAUSSIAN_C, THRESH_BINARY, 11, 2);

  Mat output4;
  threshold(noise_removed, output4, 127, 255, THRESH_BINARY);

  Mat output5;
  adaptiveThreshold(noise_removed, output5, 255, ADAPTIVE_THRESH_MEAN_C, THRESH_BINARY, 11, 2);

  Mat output6;
  adaptiveThreshold(noise_removed, output6, 255, ADAPTIVE_THRESH_GAUSSIAN_C, THRESH_BINARY, 11, 2);

  // Display results
  namedWindow("Original", WINDOW_AUTOSIZE);
  imshow("Original", original);
  
  namedWindow("Noise removed", WINDOW_AUTOSIZE);
  imshow("Noise removed", noise_removed);
  
  namedWindow("Global Thresholding (v = 127)", WINDOW_AUTOSIZE);
  imshow("Global Thresholding (v = 127)", output1);
  
  namedWindow("Adaptive Mean Thresholding", WINDOW_AUTOSIZE);
  imshow("Adaptive Mean Thresholding", output2);
  
  namedWindow("Adaptive Gaussian Thresholding", WINDOW_AUTOSIZE);
  imshow("Adaptive Gaussian Thresholding", output3);
  
  namedWindow("Global Thresholding + noise removal", WINDOW_AUTOSIZE);
  imshow("Global Thresholding + noise removal", output4);
  
  namedWindow("Adaptive Mean + noise removal", WINDOW_AUTOSIZE);
  imshow("Adaptive Gaussian Thresholding", output5);
  
  namedWindow("Adaptive Gaussian + noise removal", WINDOW_AUTOSIZE);
  imshow("Adaptive Gaussian + noise removal", output6);


  waitKey(0);

  return 0;
}

File: Adaptive Threshold for Image/AdaptiveTreshold.py
import cv2
import numpy as np
import matplotlib.pyplot as plt

# Load our new image
original_image = cv2.imread('C:\\Users\\nilfam\\Desktop\\Coding\\Futures\\Adaptive Treshold for Image\\Robot.jpg', 0)
#Noise removal
noise_removed_image = cv2.medianBlur(original_image,5)# or cv2.GaussianBlur(original_image,(5,5),0)

_,output1 = cv2.threshold(original_image,127,255,cv2.THRESH_BINARY)
output2 = cv2.adaptiveThreshold(original_image,255,cv2.ADAPTIVE_THRESH_MEAN_C, cv2.THRESH_BINARY,11,2)
output3 = cv2.adaptiveThreshold(original_image,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C, cv2.THRESH_BINARY,11,2)

_,output4 = cv2.threshold(noise_removed_image,127,255,cv2.THRESH_BINARY)
output5 = cv2.adaptiveThreshold(noise_removed_image,255,cv2.ADAPTIVE_THRESH_MEAN_C, cv2.THRESH_BINARY,11,2)
output6 = cv2.adaptiveThreshold(noise_removed_image,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C, cv2.THRESH_BINARY,11,2)


plt.figure(figsize=[15,15])
plt.subplot(331);plt.imshow(original_image, cmap='gray');plt.title("Original");
plt.subplot(332);plt.imshow(noise_removed_image, cmap='gray');plt.title("Noise removed");

plt.subplot(334);plt.imshow(output1, cmap='gray');plt.title("Global Thresholding (v = 127)");
plt.subplot(335);plt.imshow(output2, cmap='gray');plt.title("Adaptive Mean Thresholding");
plt.subplot(336);plt.imshow(output3, cmap='gray');plt.title("Adaptive Gaussian Thresholding");

plt.subplot(337);plt.imshow(output4, cmap='gray');plt.title("Global Thresholding + noise removal");
plt.subplot(338);plt.imshow(output5, cmap='gray');plt.title("Adaptive Mean + noise removal");
plt.subplot(339);plt.imshow(output6, cmap='gray');plt.title("Adaptive Gaussian + noise removal");